src/sentry/tagstore/legacy/backend.py

```diff
@@ -20,14 +20,58 @@
 
 from sentry import buffer
 from sentry.tagstore import TagKeyStatus
-from sentry.models import EventTag, GroupTagKey, GroupTagValue, TagKey, TagValue
 from sentry.tagstore.base import TagStorage
 from sentry.utils import db
 from sentry.utils.cache import cache
-from sentry.tasks.deletion import delete_tag_key
+
+from .models import EventTag, GroupTagKey, GroupTagValue, TagKey, TagValue
 
 
 class LegacyTagStorage(TagStorage):
+    def setup(self):
+        from sentry.deletions import default_manager
+        from sentry.deletions.defaults import BulkModelDeletionTask
+        from sentry.deletions.base import ModelRelation, ModelDeletionTask
+        from sentry.models import Group, Project, Event
+        from sentry.runner.commands import cleanup
+        from sentry.tasks import merge
+
+        from .deletions import TagKeyDeletionTask
+
+        default_manager.register(TagKey, TagKeyDeletionTask)
+        default_manager.register(TagValue, BulkModelDeletionTask)
+        default_manager.register(GroupTagKey, BulkModelDeletionTask)
+        default_manager.register(GroupTagValue, BulkModelDeletionTask)
+        default_manager.register(EventTag, BulkModelDeletionTask)
+
+        default_manager.add_dependencies(Group, [
+            lambda instance: ModelRelation(EventTag, {'group_id': instance.id}),
+            lambda instance: ModelRelation(GroupTagKey, {'group_id': instance.id}),
+            lambda instance: ModelRelation(GroupTagValue, {'group_id': instance.id}),
+        ])
+        default_manager.add_dependencies(Project, [
+            lambda instance: ModelRelation(TagKey, {'project_id': instance.id}),
+            lambda instance: ModelRelation(TagValue, {'project_id': instance.id}),
+            lambda instance: ModelRelation(GroupTagKey, {'project_id': instance.id}),
+            lambda instance: ModelRelation(GroupTagValue, {'project_id': instance.id}),
+        ])
+        default_manager.add_bulk_dependencies(Event, [
+            lambda instance_list: ModelRelation(EventTag,
+                                                {'event_id__in': [i.id for i in instance_list]},
+                                                ModelDeletionTask),
+        ])
+
+        cleanup.EXTRA_BULK_QUERY_DELETES += [
```

Comment:
I definitely don't like this. I'd rather add a method on `tagstore` like, `get_cleanup_models`. This is pretty gross to just be mutating a global array like this.

Reply:
I didn't love it either, but the alternative is that `deletions` and `cleanup` both have to know about `tagstore` and manually/specially reach out and ask it about this. I assumed we wanted something more general (in case another abstraction/whatever needs to do the same thing).

I'm OK with swapping it to hard-coded `tagstore.get_...` calls in those two, but I want to be sure you know that's the alternative.

(As an aside, I started making a prettier class for configuration inside of these but in the end it would be a list with an add method and it felt unnecessary to have anything more than a list.)

Reply:
I think long term, we should address this. There are a few other places, I can’t think off the top of my head, where we manually import and iterate over all apps. I believe inside of the runner for initialization is one of them. 

Might make sense longer term to introduce an apps registry similar to plugins so we can easily add hooks like this then just iterate the app registry.